### ISSUE TITLE
Using build library is becoming a productivity hindrance 

### ISSUE THREAD

**Issue by soorena776** [crossplane/crossplane#852]
### What problem are you facing?

Using [build] library across Crossplane repositories has a few nice features:
- Avoiding duplicate Makefile targets
- Standardized the CI process across build repositories

However, as we create more repositories for different stacks, using this common [build ]library also introduces some challenges:
- The library seems to be fairly opinionated. It's not clear if the benefits that were mentioned earlier outweigh the total amount of developer time that is spent in each pull request. It's also worth noting that as more and more stack repositories are created, it would be even more challenging to keep a common, agreed-upon build library for all such repositories. 
- A lot of targets in this library are not used for the purpose of the stacks. The common targets for a [stack repository] seem to be fairly simple.
- Makefiles are not easy to follow in their own. A developer who is not familiar with the library would need some time understanding the behavior of make libraries in build. In addition, there are no unit tests for the expected behavior of the targets.

### How could Crossplane help solve your problem?

I have two suggestions:
- Remove the dependency to [build] sub module in stack repositories, and create a simple Makefile in each stack repository
- If we still want to keep the library, we can retire Makefiles, in favor of a go native build system like [mage]. This will make reading and unit-testing much more easier

[build]: https://github.com/upbound/build
[mage]: https://github.com/magefile/mage
[stack repository]: https://github.com/crossplaneio/sample-stack-wordpress/blob/master/Makefile

**Comment by negz** [comment]
My thoughts on this:

* I believe there is value in maintaining a standard build tooling [golden path](https://charity.wtf/2018/12/02/software-sprawl-the-golden-path-and-scaling-teams-with-agency/) within an organisation. This allows engineers to move from project to project or team to team without having to learn a new build pipeline and process each time.
* I believe teams should be allowed to opt out of the golden path process if they feel it is hampering their productivity. The cost of this doing so is that they must maintain their own build tooling.
* I also find Makefiles hard to follow, especially when they are as complicated as ours. It is likely this is because I don't know Make as well as I know (for example) shell scripts, rather than because Make is inherently hard.
* I believe there is value in Crossplane projects "eating the dogfood" by using the build tooling we recommend Stack authors would use, which I assume probably is not Upbound's build submodule.

It's worth noting that the "golden path" pattern typically implies some people are dedicated to maintaining the golden path tooling, which is not currently the case with the Upbound build submodule. This lowers the cost for a project to opt out, because they're probably already having to implement functionality they want themselves.

**Comment by jbw976** [comment]
Thanks for sharing your thoughts on this @soorena776 and @negz.  The build submodule has worked well for a lot of projects so far since its initial investment but I can definitely empathize with the pain of having to make updates to it or to maintain it.  It's fairly complex and Makefiles have a fairly steep learning curve.

The build submodule has a ton of useful and well tested functionality that we've been relying on for awhile now around build automation, cross compilation, branching, tagging, promoting, publishing images/docs/helm, vendoring dependencies, running unit and integration tests, standardizing our toolchain, etc.  There's a lot of useful stuff in it :)

That being said, I also have the sense that it is not what we should be recommending (or forcing) stack authors to use when they are building and publishing their own stacks.  The crossplane stacks CLI family is the way forward for new stack authors, and I'm a big proponent of dogfooding.  We should be using the same tooling to build our own stacks that we are designing for new stack authors.

So we are currently in a position where the stacks CLI isn't quite ready for building and releasing full projects yet (such as our cloud provider stacks), and for expediency sake, as well as full build functionality sake, we used the build submodule to get us bootstrapped.

Probably the right investment here long term will be into the stacks CLI to get it to a place where we can run real stacks that we publish and release on a regular basis via automation.  We'll need to prioritize and find the resources to invest in that, but in the mean time we can continue using the existing build submodule that is already set up for each of the cloud provider stack repos.  Hopefully the pain is kept to a minimum if we aren't adding or updating much within the build submodule.

By the way, the current form of the stacks CLI tooling generates Makefiles to build and publish the stack.  `make` has a lot of powerful functionality for handling builds and dependencies, that's essentially what it was designed for 🤔

**Comment by displague** [comment]
I had a tin-foil hat thought that this issue approaches, so I'll park the thought here.

> That being said, I also have the sense that it is not what we should be recommending (or forcing) stack authors to use when they are building and publishing their own stacks. The crossplane stacks CLI family is the way forward for new stack authors, and I'm a big proponent of dogfooding. We should be using the same tooling to build our own stacks that we are designing for new stack authors.

Organizations with custom Stacks will want CI/CD pipelines configured to test and publish their Stacks.  This often means that their stack's build process will have access to Docker registry tokens and may have access to cloud infrastructure keys used for end-to-end tests.

With that perspective, I would be cautious, at least, of incorporating an external build process with my privileged build process.

One counter argument is that every tool in the build chain has the opportunity to run off with these special tokens, even `crossplane` and `crossplane-runtime`.

Another counter argument is that the submodule SHA is fixed on the version that the Stack owner sets and updates could be scrutinized before adoption.  However, if the build submodule pulls in dynamically updated external resources (unversioned `go get` or `go install` and `curl | sh` commands), the fixed submodule SHA defence is diminished.

**Comment by negz** [comment]
Relates to https://github.com/crossplane/crossplane/issues/1583.
